# Bank Marketing

The <b>bank-marketing.csv</b> data is related with direct marketing campaigns of a Portuguese banking institution. The marketing campaigns were based on phone calls. Often, more than one contact to the same client was required, in order to access if the product (bank term deposit) would be (or not) subscribed. The ultimate goal is to predict if the client will subscribe to a term deposit (variable y). This is a classic classification problem where the attempt is to classify between two classes - those who'll subscribe and those who won't.

Dataset reference:
- S. Moro, R. Laureano and P. Cortez. Using Data Mining for Bank Direct Marketing: An Application of the CRISP-DM Methodology. 
- In P. Novais et al. (Eds.), Proceedings of the European Simulation and Modelling Conference - ESM'2011, pp. 117-121, Guimarães, Portugal, October, 2011. EUROSIS.

#### Variable description:

- 1 age (numeric)
- 2 job : type of job (categorical: "admin.","unknown","unemployed","management","housemaid","entrepreneur","student", "blue-collar","self-employed","retired","technician","services") 
- 3 marital : marital status (categorical: "married","divorced","single"; note: "divorced" means divorced or widowed)
- 4 education (categorical: "unknown","secondary","primary","tertiary")
- 5 default: has credit in default? (binary: "yes","no")
- 6 balance: average yearly balance, in euros (numeric) 
- 7 housing: has housing loan? (binary: "yes","no")
- 8 loan: has personal loan? (binary: "yes","no")
   
#### related with the last contact of the current campaign:
- 9 contact: contact communication type (categorical: "unknown","telephone","cellular") 
- 10 day: last contact day of the month (numeric)
- 11 month: last contact month of year (categorical: "jan", "feb", "mar", ..., "nov", "dec")
- 12 duration: last contact duration, in seconds (numeric)

#### other attributes:
- 13 campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)
- 14 pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric, -1 means client was not previously contacted)
- 15 previous: number of contacts performed before this campaign and for this client (numeric)
- 16 poutcome: outcome of the previous marketing campaign (categorical: "unknown","other","failure","success")

Output variable (desired target):
- 17 y - has the client subscribed a term deposit? (binary: "yes","no")

### Read the dataset and answer the following questions.

In [2]:
import pandas as pd
df=pd.read_csv('C:\\Users\\Hang\\Documents\\Minh\\Classes\\IoT\\HW5\\bank-marketing.csv')
df

age            job  marital  education default  balance housing loan  \
0      30     unemployed  married    primary      no     1787      no   no   
1      33       services  married  secondary      no     4789     yes  yes   
2      35     management   single   tertiary      no     1350     yes   no   
3      30     management  married   tertiary      no     1476     yes  yes   
4      59    blue-collar  married  secondary      no        0     yes   no   
...   ...            ...      ...        ...     ...      ...     ...  ...   
4516   33       services  married  secondary      no     -333     yes   no   
4517   57  self-employed  married   tertiary     yes    -3313     yes  yes   
4518   57     technician  married  secondary      no      295      no   no   
4519   28    blue-collar  married  secondary      no     1137      no   no   
4520   44   entrepreneur   single   tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
0     cellular   19   oct        79         1     -1         0  unknown  no  
1     cellular   11   may       220         1    339         4  failure  no  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[4521 rows x 17 columns]

### Question

Extract all column names. Count the number of columns (using code).

In [5]:
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'balance', 'housing',
       'loan', 'contact', 'day', 'month', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'y'],
      dtype='object')

In [6]:
len(df.columns)

17

### Question

Is data in the correct format? By that we mean, do you see integers and floats where you expect them to be the case? If not, convert them into the correct format. Also make sure that all entries are non-null.

In [9]:
df.dtypes

age           int64
job          object
marital      object
education    object
default      object
balance       int64
housing      object
loan         object
contact      object
day           int64
month        object
duration      int64
campaign      int64
pdays         int64
previous      int64
poutcome     object
y            object
dtype: object

There are several columns of type Category in the data frames, so they should be converted to Category data type.

In [20]:
df['job'] = pd.Categorical(df['job'])

In [22]:
df['marital'] = pd.Categorical(df['marital'])

In [23]:
df['education'] = pd.Categorical(df['education'])

In [ ]:
df['contact'] = pd.Categorical(df['contact'])

In [ ]:
df['month'] = pd.Categorical(df['month'])

In [24]:
df['poutcome'] = pd.Categorical(df['poutcome'])

In [10]:
df.isnull().values.any()

False

In [30]:
df.isnull().sum()

age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64

### Question

Provide a general summary statistics of the entire dataset. The describe() method is what you would want to use.

In [11]:
df.describe()

age       balance          day     duration     campaign  \
count  4521.000000   4521.000000  4521.000000  4521.000000  4521.000000   
mean     41.170095   1422.657819    15.915284   263.961292     2.793630   
std      10.576211   3009.638142     8.247667   259.856633     3.109807   
min      19.000000  -3313.000000     1.000000     4.000000     1.000000   
25%      33.000000     69.000000     9.000000   104.000000     1.000000   
50%      39.000000    444.000000    16.000000   185.000000     2.000000   
75%      49.000000   1480.000000    21.000000   329.000000     3.000000   
max      87.000000  71188.000000    31.000000  3025.000000    50.000000   

             pdays     previous  
count  4521.000000  4521.000000  
mean     39.766645     0.542579  
std     100.121124     1.693562  
min      -1.000000     0.000000  
25%      -1.000000     0.000000  
50%      -1.000000     0.000000  
75%      -1.000000     0.000000  
max     871.000000    25.000000

### Question

The data type of columns like job, married, education etc. is called categorical data. List all the different categories in the job column.

In [25]:
df['job'].cat.categories



Index(['admin', 'blue-collar', 'entrepreneur', 'housemaid', 'management',
       'retired', 'self-employed', 'services', 'student', 'technician',
       'unemployed', 'unknown'],
      dtype='object')

### Question

In one line of code, provide a count of the number of people who were unemployed and owned a home.

In [29]:
len(df[(df['job']=='unemployed') & (df['housing'] == 'yes')])

58

### Question

What is the education level of a typical blue-collar worker? Explore value_counts() method.

In [38]:
blueCollar=df[df['job']=='blue-collar']
blueCollar['education'].value_counts(normalize=True).head(1)

secondary    0.553911
Name: education, dtype: float64

55% of the blue-collar workers have secondary education.  Thus, the typical blue-collar worker has a secondary education.

### Question

How many who are unemployed have an outstanding loan? Is that percentage more than that of the employed ones? 

In [45]:
totalUnemployed=df[df['job']=='unemployed']
totalEmployed=df[df['job'] != 'unemployed']
unemployedWithBalance = totalUnemployed[totalUnemployed['balance']>0]
employedWithBalance = totalEmployed[totalEmployed['balance']>0]
len(unemployedWithBalance)

102

In [46]:
len(unemployedWithBalance)/len(totalUnemployed)

0.796875

In [49]:
len(employedWithBalance)/len(totalEmployed)

0.8413384930571364

Thus 79.7% of unemployed has outstanding loan while 84.1% of employed has outstanding loan.

### Question

What percent of clients subscribed to the term deposit (column y)? 

In [51]:
df['y'].value_counts(normalize=True)

no     0.88476
yes    0.11524
Name: y, dtype: float64

11.5% of clients subscribed to the term deposit.

### Question

What percent of married clients subscribed to the term deposit? Is that more or less than that for single folks? 

In [52]:
df[df['marital']=='married']['y'].value_counts(normalize=True)

no     0.900965
yes    0.099035
Name: y, dtype: float64

In [53]:
df[df['marital']!='married']['y'].value_counts(normalize=True)

no     0.858469
yes    0.141531
Name: y, dtype: float64

Based on the above info, 9.9% married people subscribed to the term deposit, which is less than 14% of single people who do.

### Question

<p>Ask an interesting question of this data set and provide a solution to answer that.</p>
<p>The goal is to help teach your fellow students all possible questions we can collectively ask of this dataset. Your question should be as clear as possible and as short as possible. Try to avoid asking questions that are too trivial or obvious.</p>

<p>This is a bonus point question. The only way not to get full points are if you do the following:
<ul>
<li>You do not perform this task.
<li>You do not include a solution. 
</ul>
</p>
<p>If the solution you provide is incorrect, you will still receive full points. But you must make an honest effort to get it right.
</p>

Write your question here.

**Question**: For all people with tertiary education level, what is the percentage of those who have been contacted in the last 10 days <br> but have not signed up for the term deposit yet?<br> Of these undecided people, how many are unemployed? 

In [63]:
tertiary = df[df['education']=='tertiary']
tertContactLast10Days = tertiary[(tertiary['pdays'] <= 10) & (tertiary['pdays'] != -1)]
tertContactLast10DaysButNotSign = tertContactLast10Days[tertContactLast10Days['y']=='no']

percentOfTertiaryStillNotSign = len(tertContactLast10DaysButNotSign) / len(tertiary)
percentOfTertiaryStillNotSign

0.0044444444444444444

So the first answer is that 0.4% of the people with tertiary education level have been contacted in the last 10 days but still have not signed up for the term deposit yet.

In [66]:
len(tertContactLast10DaysButNotSign[tertContactLast10DaysButNotSign['job']=='unemployed'])

0

Thus, for the undecided people from the result set above, no one is unemployed. So it is not a money factor that blocks them from subscribing the term deposit.